### PR TITLE
#1731 Read the Docs configuration file v2

### LINKED ISSUE #1731: Unable to generate documentation on ReadTheDocs
## Expected Behavior / New Feature
We would expect the documentation to be generated on a regular basis, to avoid misunderstandings. 😸 

## Actual Behavior / Motivation for New Feature
The documentation generation process fails at the moment:
[Build #22151302 | Ocelot | Read the Docs](https://readthedocs.org/projects/ocelot/builds/22151302/)

The reason is that Read the Docs has stopped supporting docs generation without the `.readthedocs.yaml` config file:
[Migrate your project to .readthedocs.yaml configuration file v2 — Read the Docs Blog](https://blog.readthedocs.com/migrate-configuration-v2/)

## Solution
A simple solution is given in the PR. By adding a config file to the solution's root folder the docs are generated.
See demo account by @ggnaegi: https://ocelot-test.readthedocs.io/en/bug-fix_documentation_generation/

### PR DESCRIPTION
**Pull request by ggnaegi**
## Fixes #1731 
- #1731 

## Proposed Changes
  - add `.readthedocs.yaml` file to the root folder.
  - update to the latest sphinx version, update `conf.py`

Blog article: [Migrate your project to .readthedocs.yaml configuration file v2 — Read the Docs Blog](https://blog.readthedocs.com/migrate-configuration-v2/)

I just tried myself
https://ocelot-test.readthedocs.io/en/latest/features/authentication.html

### CODE REVIEW

**Comment by ggnaegi** [comment]
@raman-m Ok, done, you can merge

**Comment by raman-m** [comment]
> @raman-m Ok, done, you can merge

Not yet! 😉

File: .readthedocs.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  tools:
+    python: "3.11"
+    # You can also specify other tool versions:
+    # nodejs: "19"
+    # rust: "1.64"
+    # golang: "1.19"
+
+# Build documentation in the "docs/" directory with Sphinx
+sphinx:
+   configuration: docs/conf.py
```
Comment (raman-m) [inline .readthedocs.yaml:20]: We have to review and update this file!

**Review by raman-m — CHANGES_REQUESTED** [review]
Technically adding mentioned **.yaml** file will resolve the problem of building, but let's update the `docs/conf.py` file too. I find the file very outdated.

**Comment by raman-m** [comment]
### First,
The copyright message looks strange: https://ocelot-test.readthedocs.io/en/latest/index.html
Copyright is located at the bottom of each page!
image
It is 2023 year now and project belongs to ThreeMammals org.
The line: [docs/conf.py | Line 62](../blob/develop/docs/conf.py#L62)

### Second,
Version of docs is 1.0.0 !!! 🤣 
image
So, HTML doc title contains this info:
image
But actually version should be 20.0 because of current release 20.0
So, it is hardcoded now in config:
```python
# The short X.Y version.
version = '1.0.0'
```
The line: [docs/conf.py | Line 70](../blob/develop/docs/conf.py#L70)

### Third,
Release info is also wrong, it is 1.0.0 now. But I didn't find how it is reflected to documentation files.
```python
# The full version, including alpha/beta/rc tags.
release = '1.0.0'
```
The line: [docs/conf.py | Line 72](../blob/develop/docs/conf.py#L72)


### I propose to fix these issues. Hardcode for this hotfix is OK. But in future we have to write dynamic calculation of copyright year, version and release variables.

**Comment by raman-m** [comment]
### Important blog article
[Doctools without configuration file (conf.py / mkdocs.yml) are deprecated — Read the Docs Blog](https://blog.readthedocs.com/doctool-without-configuration-file/) on 12 July 2023

File: .readthedocs.yaml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+# Optionally build your docs in additional formats such as PDF and ePub
+# formats:
+#    - pdf
+#    - epub
```
Comment (raman-m) [inline .readthedocs.yaml:25]: We can try to add these formats. Otherwise we will have HTML docs only, right?

File: .readthedocs.yaml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+# Optional but recommended, declare the Python requirements required
+# to build your documentation
+# See https://docs.readthedocs.io/en/stable/guides/reproducible-builds.html
+# python:
+#    install:
+#    - requirements: docs/requirements.txt
```
Comment (raman-m) [inline .readthedocs.yaml:32]: And it seems wee could add this file too!
More: https://docs.readthedocs.io/en/stable/guides/reproducible-builds.html

**Review by raman-m — CHANGES_REQUESTED** [review]
Make corrections please!

File: .readthedocs.yaml (hunk 1 above)
Comment (ggnaegi) [reply .readthedocs.yaml:20]: @raman-m Ok, it was indeed very outdated, I have now updated the conf.py to the latest version (7.2.6)

File: .readthedocs.yaml (hunk 3 above)
Comment (ggnaegi) [reply .readthedocs.yaml]: Ok, done

**Comment by ggnaegi** [comment]
@raman-m So, you can check it here: https://ocelot-test.readthedocs.io/en/bug-fix_documentation_generation

I have added the requirements.txt file, made some improvements in the doc (fixing documentation compilation warnings), fixed the copyrights and versions.

File: .readthedocs.yaml (hunk 2 above)
Comment (ggnaegi) [reply .readthedocs.yaml]: yes

**Comment by raman-m** [comment]
> (fixing documentation compilation warnings)

What are you talking about?

**Comment by raman-m** [comment]
> @raman-m So, you can check it here: https://ocelot-test.readthedocs.io/en/bug-fix_documentation_generation
> 
> I have added the requirements.txt file, made some improvements in the doc (fixing documentation compilation warnings), fixed the copyrights and versions.

Thank you!

**Comment by ggnaegi** [comment]
> > (fixing documentation compilation warnings)
> 
> What are you talking about?

When building the documentation locally...
image

That's what I fixed.
```powershell
D:\dev\repositories\sphinx_test\building/releaseprocess.rst:19: WARNING: Enumerated list ends without a blank line; unexpected unindent.
D:\dev\repositories\sphinx_test\features/authentication.rst:173: WARNING: Title underline too short.

Allowed Scopes
^^^^^^^^^^^^^
D:\dev\repositories\sphinx_test\features/authentication.rst:173: WARNING: Title underline too short.

Allowed Scopes
^^^^^^^^^^^^^
D:\dev\repositories\sphinx_test\features/claimstransformation.rst:18: WARNING: Title underline too short.

Claims to Claims Transformation
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
D:\dev\repositories\sphinx_test\features/claimstransformation.rst:45: WARNING: Title underline too short.

Claims to Query String Parameters Transformation
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
D:\dev\repositories\sphinx_test\features/claimstransformation.rst:45: WARNING: Title underline too short.

Claims to Query String Parameters Transformation
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
D:\dev\repositories\sphinx_test\features/kubernetes.rst:31: WARNING: Explicit markup ends without a blank line; unexpected unindent.
D:\dev\repositories\sphinx_test\features/kubernetes.rst:57: WARNING: Explicit markup ends without a blank line; unexpected unindent.
D:\dev\repositories\sphinx_test\features/loadbalancer.rst:105: WARNING: Title underline too short.

Custom Load Balancers
^^^^^^^^^^^^^^^^^^^^
D:\dev\repositories\sphinx_test\features/loadbalancer.rst:105: WARNING: Title underline too short.

Custom Load Balancers
^^^^^^^^^^^^^^^^^^^^
D:\dev\repositories\sphinx_test\features/tracing.rst:2: WARNING: Duplicate explicit target name: "here".
D:\dev\repositories\sphinx_test\features/websockets.rst:76: WARNING: Explicit markup ends without a blank line; unexpected unindent.
```

**Review by raman-m — APPROVED** [review]

**Comment by raman-m** [comment]
> @ggnaegi commented [on Oct 15](#issuecomment-1763481237)

My local build results:
image
I have 3 warnings! 😄

**Comment by ggnaegi** [comment]
> > @ggnaegi commented [on Oct 15](#issuecomment-1763481237)
> 
> My local build results:
> image
> I have 3 warnings! 😄 

@raman-m Yeah, rebuild it again and it's going to be fine, only the static folder 😁

My concern is now the following: what about main branch?

**Comment by raman-m** [comment]
> My concern is now the following: what about main branch?

Tomorrow I will create a PR to main from this release branch. The PR must be approved by Raynald. And then, we will get new release. And will check actual Docs build here: https://ocelot.readthedocs.io/
I will keep you updated...

**Pull request merged**